feat: add unary function database in `math/special/data`

gunjjoshi opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: na
  - task: lint_package_json status: na
  - task: lint_repl_help status: na
  - task: lint_javascript_src status: na
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: na
  - task: lint_javascript_tests status: na
  - task: lint_javascript_benchmarks status: na
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: na
  - task: lint_c_examples status: na
  - task: lint_c_benchmarks status: na
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: na
  - task: lint_typescript_tests status: na
  - task: lint_license_headers status: passed ---

Resolves None.

## Description

> What is the purpose of this pull request?

This pull request:

-   adds the unary function database in `math/special/data`, as per https://github.com/stdlib-js/stdlib/pull/7773/files#r2337722847.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves None.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

gunjjoshi committed:
feat: add unary function database

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi committed:
docs: add data directory in package.json

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/math/special/data/unary_function_database.json:
```diff
@@ -0,0 +1,3386 @@
+      "complex32"
+    ],
+    "scalar_kernels": {
+      "int8": "@stdlib/math/base/special/labs",
+      "int16": "@stdlib/math/base/special/labs",
```
In thinking about this, if we don't have dedicated kernels for particular dtypes (except for "generic"), we shouldn't include them in the list of scalar kernels.

In the generation script, I believe we try to find the nearest "compatible" scalar kernel based on type promotion, correct? In which case, for `int16` and `int8` here, we'd infer that `labs` should be used.

kgryte commented on lib/node_modules/@stdlib/math/special/data/unary_function_database.json:
```diff
@@ -0,0 +1,3386 @@
+      "complex32"
+    ],
+    "scalar_kernels": {
+      "int8": "@stdlib/math/base/special/abs2f",
+      "int16": "@stdlib/math/base/special/abs2f",
+      "int32": "@stdlib/math/base/special/abs2f",
+      "uint8": "@stdlib/math/base/special/abs2f",
+      "uint16": "@stdlib/math/base/special/abs2f",
+      "uint32": "@stdlib/math/base/special/abs2f",
```
```suggestion
```

We should remove these, as `abs2f` is not universally appropriate. It depends on the output dtype what kernel should be used. That applies here and for other functions in this file.

kgryte commented on lib/node_modules/@stdlib/math/special/data/unary_function_database.json:
```diff
@@ -0,0 +1,3386 @@
+      "generic": "@stdlib/math/base/special/absgammalnf"
+    },
+    "policies": {
+      "output": "real_floating_point",
```
```suggestion
      "output": "real_floating_point_and_generic",
```

kgryte commented on lib/node_modules/@stdlib/math/special/data/unary_function_database.json:
```diff
@@ -0,0 +1,3386 @@
+      "generic": "@stdlib/math/base/special/acos"
+    },
+    "policies": {
+      "output": "real_floating_point",
```
```suggestion
      "output": "real_floating_point_and_generic",
```

Anywhere where `generic` is an accepted input dtype, it should also be an accepted output dtype.

kgryte commented on lib/node_modules/@stdlib/math/special/data/unary_function_database.json:
```diff
@@ -0,0 +1,3386 @@
+      "uint8": "@stdlib/math/base/special/binet",
+      "uint16": "@stdlib/math/base/special/binet",
+      "uint32": "@stdlib/math/base/special/binet",
+      "float32": "@stdlib/math/base/special/binet",
```
If we don't have a `float32` kernel, we should not include.

kgryte reviewed (changes requested):
Thanks for putting together an initial draft, @gunjjoshi. Left some initial comments for clean-up.

gunjjoshi committed:
refactor: exclude unwanted dtypes, add generic to output poicy

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi commented:
Thanks for having a look, @kgryte. I have updated the PR. Now,
- We have only specialized scalar kernels mentioned, i.e., the `dtype` is listed only if it has a dedicated scalar kernel.
- The `output` field inside `policies` now includes `generic` too, for all functions that contain `generic` in input dtypes.

kgryte commented on lib/node_modules/@stdlib/math/special/data/unary_function_database.json:
```diff
@@ -0,0 +1,2610 @@
+      "casting": "none"
+    }
+  },
+  "absgammalnf": {
```
```suggestion
  "absgammaln": {
```

kgryte commented on lib/node_modules/@stdlib/math/special/data/unary_function_database.json:
```diff
@@ -0,0 +1,2610 @@
+    "scalar_kernels": {
+      "float32": "@stdlib/math/base/special/absgammalnf",
+      "float64": "@stdlib/math/base/special/absgammalnf",
+      "generic": "@stdlib/math/base/special/absgammalnf"
```
```suggestion
      "float64": "@stdlib/math/base/special/gammaln",
      "generic": "@stdlib/math/base/special/gammaln"
```

kgryte commented on lib/node_modules/@stdlib/math/special/data/unary_function_database.json:
```diff
@@ -0,0 +1,2610 @@
+  },
+  "bernoulli": {
+    "input_dtypes": "real_and_generic",
+    "output_dtypes": "real_and_generic",
+    "excluded_dtypes": [
+      "float16",
+      "uint8c",
+      "int64",
+      "uint64"
+    ],
+    "scalar_kernels": {
+      "float32": "@stdlib/math/base/special/bernoullif",
+      "float64": "@stdlib/math/base/special/bernoulli",
+      "generic": "@stdlib/math/base/special/bernoulli"
+    },
+    "policies": {
+      "output": "real_and_generic",
```
```suggestion
    "output_dtypes": "real_floating_point_and_generic",
    "excluded_dtypes": [
      "float16",
      "uint8c",
      "int64",
      "uint64"
    ],
    "scalar_kernels": {
      "float32": "@stdlib/math/base/special/bernoullif",
      "float64": "@stdlib/math/base/special/bernoulli",
      "generic": "@stdlib/math/base/special/bernoulli"
    },
    "policies": {
      "output": "real_floating_point_and_generic",
```

kgryte commented on lib/node_modules/@stdlib/math/special/data/unary_function_database.json:
```diff
@@ -0,0 +1,2610 @@
+  },
+  "binet": {
+    "input_dtypes": "real_and_generic",
+    "output_dtypes": "real_and_generic",
+    "excluded_dtypes": [
+      "float16",
+      "uint8c",
+      "int64",
+      "uint64"
+    ],
+    "scalar_kernels": {
+      "float64": "@stdlib/math/base/special/binet",
+      "generic": "@stdlib/math/base/special/binet"
+    },
+    "policies": {
+      "output": "real_and_generic",
```
```suggestion
    "output_dtypes": "real_floating_point_and_generic",
    "excluded_dtypes": [
      "float16",
      "uint8c",
      "int64",
      "uint64"
    ],
    "scalar_kernels": {
      "float64": "@stdlib/math/base/special/binet",
      "generic": "@stdlib/math/base/special/binet"
    },
    "policies": {
      "output": "real_floating_point_and_generic",
```

kgryte commented on lib/node_modules/@stdlib/math/special/data/unary_function_database.json:
```diff
@@ -0,0 +1,2610 @@
+      "uint64"
+    ],
+    "scalar_kernels": {
+      "float32": "@stdlib/math/base/special/erf",
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/math/special/data/unary_function_database.json:
```diff
@@ -0,0 +1,2610 @@
+    }
+  },
+  "exp": {
+    "input_dtypes": "numeric_and_generic",
+    "output_dtypes": "numeric_and_generic",
+    "excluded_dtypes": [
+      "float16",
+      "uint8c",
+      "int64",
+      "uint64",
+      "complex32"
+    ],
+    "scalar_kernels": {
+      "float64": "@stdlib/math/base/special/exp",
+      "complex128": "@stdlib/math/base/special/cexp",
+      "generic": "@stdlib/math/base/special/exp"
+    },
+    "policies": {
+      "output": "numeric_and_generic",
```
```suggestion
    "input_dtypes": "numeric_and_generic",
    "output_dtypes": "floating_point_and_generic",
    "excluded_dtypes": [
      "float16",
      "uint8c",
      "int64",
      "uint64",
      "complex32"
    ],
    "scalar_kernels": {
      "float64": "@stdlib/math/base/special/exp",
      "complex128": "@stdlib/math/base/special/cexp",
      "generic": "@stdlib/math/base/special/exp"
    },
    "policies": {
      "output": "floating_point_and_generic",
```

kgryte commented on lib/node_modules/@stdlib/math/special/data/unary_function_database.json:
```diff
@@ -0,0 +1,2610 @@
+  },
+  "fibonacci": {
+    "input_dtypes": "real_and_generic",
+    "output_dtypes": "real_and_generic",
+    "excluded_dtypes": [
+      "float16",
+      "uint8c",
+      "int64",
+      "uint64"
+    ],
+    "scalar_kernels": {
+      "float32": "@stdlib/math/base/special/fibonaccif",
+      "float64": "@stdlib/math/base/special/fibonacci",
+      "generic": "@stdlib/math/base/special/fibonacci"
+    },
+    "policies": {
+      "output": "real_and_generic",
```
```suggestion
    "output_dtypes": "real_floating_point_and_generic",
    "excluded_dtypes": [
      "float16",
      "uint8c",
      "int64",
      "uint64"
    ],
    "scalar_kernels": {
      "float32": "@stdlib/math/base/special/fibonaccif",
      "float64": "@stdlib/math/base/special/fibonacci",
      "generic": "@stdlib/math/base/special/fibonacci"
    },
    "policies": {
      "output": "real_floating_point_and_generic",
```

kgryte commented on lib/node_modules/@stdlib/math/special/data/unary_function_database.json:
```diff
@@ -0,0 +1,2610 @@
+  },
+  "fibonacci-index": {
+    "input_dtypes": "real_and_generic",
+    "output_dtypes": "real_and_generic",
+    "excluded_dtypes": [
+      "float16",
+      "uint8c",
+      "int64",
+      "uint64"
+    ],
+    "scalar_kernels": {
+      "float32": "@stdlib/math/base/special/fibonacci-indexf",
+      "float64": "@stdlib/math/base/special/fibonacci-index",
+      "generic": "@stdlib/math/base/special/fibonacci-index"
+    },
+    "policies": {
+      "output": "real_and_generic",
```
```suggestion
    "output_dtypes": "real_floating_point_and_generic",
    "excluded_dtypes": [
      "float16",
      "uint8c",
      "int64",
      "uint64"
    ],
    "scalar_kernels": {
      "float32": "@stdlib/math/base/special/fibonacci-indexf",
      "float64": "@stdlib/math/base/special/fibonacci-index",
      "generic": "@stdlib/math/base/special/fibonacci-index"
    },
    "policies": {
      "output": "real_floating_point_and_generic",
```

kgryte commented on lib/node_modules/@stdlib/math/special/data/unary_function_database.json:
```diff
@@ -0,0 +1,2610 @@
+      "casting": "none"
+    }
+  },
+  "frexp": {
+    "input_dtypes": "real_and_generic",
+    "output_dtypes": "numeric_and_generic",
+    "policies": {
+      "output": "numeric_and_generic",
+      "casting": "none"
+    },
+    "excluded_dtypes": [
+      "float16",
+      "uint8c",
+      "int64",
+      "uint64"
+    ],
+    "scalar_kernels": {
+      "float32": "@stdlib/math/base/special/frexpf",
+      "float64": "@stdlib/math/base/special/frexp",
+      "generic": "@stdlib/math/base/special/frexp"
+    }
+  },
```
```suggestion
```

`frexp` is not an ordinary unary function, as it outputs two values, not one.

kgryte commented on lib/node_modules/@stdlib/math/special/data/unary_function_database.json:
```diff
@@ -0,0 +1,2610 @@
+      "casting": "none"
+    }
+  },
+  "gammaln": {
+    "input_dtypes": "real_and_generic",
+    "output_dtypes": "real_floating_point_and_generic",
+    "excluded_dtypes": [
+      "float16",
+      "uint8c",
+      "int64",
+      "uint64"
+    ],
+    "scalar_kernels": {
+      "float64": "@stdlib/math/base/special/gammaln",
+      "generic": "@stdlib/math/base/special/gammaln"
+    },
+    "policies": {
+      "output": "real_floating_point_and_generic",
+      "casting": "none"
+    }
+  },
```
```suggestion
```

This is covered by `absgammaln` above.

kgryte commented on lib/node_modules/@stdlib/math/special/data/unary_function_database.json:
```diff
@@ -0,0 +1,2610 @@
+  },
+  "inv": {
+    "input_dtypes": "numeric_and_generic",
+    "output_dtypes": "numeric_and_generic",
+    "excluded_dtypes": [
+      "float16",
+      "uint8c",
+      "int64",
+      "uint64",
+      "complex32"
+    ],
+    "scalar_kernels": {
+      "float32": "@stdlib/math/base/special/invf",
+      "float64": "@stdlib/math/base/special/inv",
+      "complex64": "@stdlib/math/base/special/cinvf",
+      "complex128": "@stdlib/math/base/special/cinv",
+      "generic": "@stdlib/math/base/special/inv"
+    },
+    "policies": {
+      "output": "numeric_and_generic",
```
```suggestion
    "output_dtypes": "floating_point_and_generic",
    "excluded_dtypes": [
      "float16",
      "uint8c",
      "int64",
      "uint64",
      "complex32"
    ],
    "scalar_kernels": {
      "float32": "@stdlib/math/base/special/invf",
      "float64": "@stdlib/math/base/special/inv",
      "complex64": "@stdlib/math/base/special/cinvf",
      "complex128": "@stdlib/math/base/special/cinv",
      "generic": "@stdlib/math/base/special/inv"
    },
    "policies": {
      "output": "floating_point_and_generic",
```

kgryte commented on lib/node_modules/@stdlib/math/special/data/unary_function_database.json:
```diff
@@ -0,0 +1,2610 @@
+  },
+  "lucas": {
+    "input_dtypes": "real_and_generic",
+    "output_dtypes": "real_and_generic",
+    "excluded_dtypes": [
+      "float16",
+      "uint8c",
+      "int64",
+      "uint64"
+    ],
+    "scalar_kernels": {
+      "float32": "@stdlib/math/base/special/lucasf",
+      "float64": "@stdlib/math/base/special/lucas",
+      "generic": "@stdlib/math/base/special/lucas"
+    },
+    "policies": {
+      "output": "real_and_generic",
```
```suggestion
    "output_dtypes": "real_floating_point_and_generic",
    "excluded_dtypes": [
      "float16",
      "uint8c",
      "int64",
      "uint64"
    ],
    "scalar_kernels": {
      "float32": "@stdlib/math/base/special/lucasf",
      "float64": "@stdlib/math/base/special/lucas",
      "generic": "@stdlib/math/base/special/lucas"
    },
    "policies": {
      "output": "real_floating_point_and_generic",
```

kgryte commented on lib/node_modules/@stdlib/math/special/data/unary_function_database.json:
```diff
@@ -0,0 +1,2610 @@
+  },
+  "negafibonacci": {
+    "input_dtypes": "real_and_generic",
+    "output_dtypes": "real_and_generic",
+    "excluded_dtypes": [
+      "float16",
+      "uint8c",
+      "int64",
+      "uint64"
+    ],
+    "scalar_kernels": {
+      "float32": "@stdlib/math/base/special/negafibonaccif",
+      "float64": "@stdlib/math/base/special/negafibonacci",
+      "generic": "@stdlib/math/base/special/negafibonacci"
+    },
+    "policies": {
+      "output": "real_and_generic",
```
```suggestion
    "output_dtypes": "real_floating_point_and_generic",
    "excluded_dtypes": [
      "float16",
      "uint8c",
      "int64",
      "uint64"
    ],
    "scalar_kernels": {
      "float32": "@stdlib/math/base/special/negafibonaccif",
      "float64": "@stdlib/math/base/special/negafibonacci",
      "generic": "@stdlib/math/base/special/negafibonacci"
    },
    "policies": {
      "output": "real_floating_point_and_generic",
```

kgryte commented on lib/node_modules/@stdlib/math/special/data/unary_function_database.json:
```diff
@@ -0,0 +1,2610 @@
+  },
+  "negalucas": {
+    "input_dtypes": "real_and_generic",
+    "output_dtypes": "real_and_generic",
+    "excluded_dtypes": [
+      "float16",
+      "uint8c",
+      "int64",
+      "uint64"
+    ],
+    "scalar_kernels": {
+      "float32": "@stdlib/math/base/special/negalucasf",
+      "float64": "@stdlib/math/base/special/negalucas",
+      "generic": "@stdlib/math/base/special/negalucas"
+    },
+    "policies": {
+      "output": "real_and_generic",
```
```suggestion
    "output_dtypes": "real_floating_point_and_generic",
    "excluded_dtypes": [
      "float16",
      "uint8c",
      "int64",
      "uint64"
    ],
    "scalar_kernels": {
      "float32": "@stdlib/math/base/special/negalucasf",
      "float64": "@stdlib/math/base/special/negalucas",
      "generic": "@stdlib/math/base/special/negalucas"
    },
    "policies": {
      "output": "real_floating_point_and_generic",
```

kgryte commented on lib/node_modules/@stdlib/math/special/data/unary_function_database.json:
```diff
@@ -0,0 +1,2610 @@
+  },
+  "nonfibonacci": {
+    "input_dtypes": "real_and_generic",
+    "output_dtypes": "real_and_generic",
+    "excluded_dtypes": [
+      "float16",
+      "uint8c",
+      "int64",
+      "uint64"
+    ],
+    "scalar_kernels": {
+      "float32": "@stdlib/math/base/special/nonfibonaccif",
+      "float64": "@stdlib/math/base/special/nonfibonacci",
+      "generic": "@stdlib/math/base/special/nonfibonacci"
+    },
+    "policies": {
+      "output": "real_and_generic",
```
```suggestion
    "output_dtypes": "real_floating_point_and_generic",
    "excluded_dtypes": [
      "float16",
      "uint8c",
      "int64",
      "uint64"
    ],
    "scalar_kernels": {
      "float32": "@stdlib/math/base/special/nonfibonaccif",
      "float64": "@stdlib/math/base/special/nonfibonacci",
      "generic": "@stdlib/math/base/special/nonfibonacci"
    },
    "policies": {
      "output": "real_floating_point_and_generic",
```

kgryte commented on lib/node_modules/@stdlib/math/special/data/unary_function_database.json:
```diff
@@ -0,0 +1,2610 @@
+  },
+  "ramp": {
+    "input_dtypes": "real_and_generic",
+    "output_dtypes": "real_floating_point_and_generic",
+    "excluded_dtypes": [
+      "float16",
+      "uint8c",
+      "int64",
+      "uint64"
+    ],
+    "scalar_kernels": {
+      "float32": "@stdlib/math/base/special/rampf",
+      "float64": "@stdlib/math/base/special/ramp",
+      "generic": "@stdlib/math/base/special/ramp"
+    },
+    "policies": {
+      "output": "real_floating_point_and_generic",
```
```suggestion
    "output_dtypes": "real_and_generic",
    "excluded_dtypes": [
      "float16",
      "uint8c",
      "int64",
      "uint64"
    ],
    "scalar_kernels": {
      "float32": "@stdlib/math/base/special/rampf",
      "float64": "@stdlib/math/base/special/ramp",
      "generic": "@stdlib/math/base/special/ramp"
    },
    "policies": {
      "output": "real_and_generic",
```

kgryte commented on lib/node_modules/@stdlib/math/special/data/unary_function_database.json:
```diff
@@ -0,0 +1,2610 @@
+      "generic": "@stdlib/math/base/special/tribonacci"
+    },
+    "policies": {
+      "output": "real_and_generic",
```
```suggestion
    "output_dtypes": "real_floating_point_and_generic",
    "excluded_dtypes": [
      "float16",
      "uint8c",
      "int64",
      "uint64"
    ],
    "scalar_kernels": {
      "float32": "@stdlib/math/base/special/tribonaccif",
      "float64": "@stdlib/math/base/special/tribonacci",
      "generic": "@stdlib/math/base/special/tribonacci"
    },
    "policies": {
      "output": "real_floating_point_and_generic",
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM after some minor clean-up. Let's go ahead and get this in, and we can make changes in follow-up PRs/commits.
